File: 3739-count-subarrays-with-majority-element-ii/solution.cpp
#include <vector>
#include <algorithm>

using namespace std;

class FenwickTree {
    int n;
    vector<int> tree;

public:
    FenwickTree(int n) : n(n), tree(n + 1, 0) {}

    void add(int i, int delta) {
        for (; i <= n; i += i & -i) tree[i] += delta;
    }

    int query(int i) {
        int sum = 0;
        for (; i > 0; i -= i & -i) sum += tree[i];
        return sum;
    }
};

class Solution {
public:
    long long countSubarrays(vector<int>& nums, int target) {
        int n = nums.size();
        // Prefixes range from -n to n. Offset by n+1 to make them 1-indexed.
        vector<int> p(n + 1, 0);
        for (int i = 0; i < n; ++i) {
            p[i + 1] = p[i] + (nums[i] == target ? 1 : -1);
        }

        int offset = n + 1;
        int max_val = 2 * n + 1;
        FenwickTree ft(max_val);
        long long count = 0;

        for (int val : p) {
            // Count how many previous prefix sums are strictly less than current
            count += ft.query(val + offset - 1);
            ft.add(val + offset, 1);
        }

        return count;
    }
};

File: 3739-count-subarrays-with-majority-element-ii/solution.py
class FenwickTree:
    def __init__(self, n):
        self.tree = [0] * (n + 1)

    def add(self, i, delta):
        while i < len(self.tree):
            self.tree[i] += delta
            i += i & (-i)

    def query(self, i):
        s = 0
        while i > 0:
            s += self.tree[i]
            i -= i & (-i)
        return s

class Solution:
    def countSubarrays(self, nums: list[int], target: int) -> int:
        n = len(nums)
        # Prefix sum P[i] = count(target) - count(not target)
        # P[j+1] - P[i] > 0  => P[j+1] > P[i]
        p = [0] * (n + 1)
        for i in range(n):
            p[i + 1] = p[i] + (1 if nums[i] == target else -1)
        
        # Offset to handle negative prefix sums. Range is [-n, n]
        offset = n + 1
        ft = FenwickTree(2 * n + 2)
        count = 0
        
        for val in p:
            # Add offset to make values strictly positive for Fenwick Tree
            # Query count of all previous p values < current val
            count += ft.query(val + offset - 1)
            ft.add(val + offset, 1)
            
        return count